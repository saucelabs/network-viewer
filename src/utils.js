export const getUrlInfo = (url) => {
  const urlInfo = new URL(url);
  const pathSplit = urlInfo.pathname.split('/');
  const fileName = pathSplit[pathSplit.length - 1].trim()
    ? pathSplit[pathSplit.length - 1] : pathSplit[pathSplit.length - 2];

  return {
    domain: urlInfo.host,
    filename: fileName + urlInfo.search,
    url: urlInfo.href,
  };
};

export const parseSize = ({ headers, content }) => {
  if (content && content.size) {
    return Number((content.size / 1024).toFixed(2));
  }

  const contentInfo = headers.find(({ name }) => ['content-length', 'Content-Length'].includes(name));
  if (!contentInfo) {
    return 0;
  }

  return Number((contentInfo.value / 1024).toFixed(2));
};

export const getContentType = (headers) => {
  const contentInfo = headers.find(({ name }) => ['content-type', 'Content-Type'].includes(name));
  if (!contentInfo) {
    return '';
  }
  const type = contentInfo.value.split(';')[0].split('/');
  return type.length > 1 ? type[1] : type[0];
};

export const getTimings = ({ startedDateTime, timings }, firstEntryTime) => ({
  ...timings,
  startTime: new Date(startedDateTime).getTime() - new Date(firstEntryTime).getTime(),
});

export const prepareViewerData = (entries) => {
  const firstEntryTime = entries[0].startedDateTime;
  const lastEntryTime = entries[entries.length - 1].startedDateTime;
  const data = entries
    .filter(entry => entry.response)
    .map((entry, index) => ({
      index,
      status: entry.response.status,
      method: entry.request.method,
      size: parseSize(entry.response),
      startedDateTime: entry.startedDateTime,
      type: getContentType(entry.response.headers),
      timings: getTimings(entry, firstEntryTime),
      ...getUrlInfo(entry.request.url),
    }));

  const totalNetworkTime = new Date(lastEntryTime).getTime()
    - new Date(firstEntryTime).getTime()
    + data[data.length - 1].timings.receive;
  return {
    totalNetworkTime,
    data,
  };
};

export const sortBy = (data, key, isAsc = true) => data.sort((prev, next) => {
  if (prev[key] < next[key]) {
    return isAsc ? -1 : 1;
  }
  if (prev[key] > next[key]) {
    return isAsc ? 1 : 1;
  }
  return 0;
});

export const filterData = (data, filter) => {
  const { search, key, value } = filter;
  const trimmedSearch = search.trim();

  return !trimmedSearch && !key
    ? data
    : data.filter((info) => {
      const isSearchMatched = trimmedSearch ? info.filename.includes(trimmedSearch) : true;
      const isFilterMatched = key ? info[key] === value : true;
      return isSearchMatched && isFilterMatched;
    });
};

export const actionsWrapper = (actions = {}) => (dispatch, state) => Object.keys(actions)
  .reduce((modifiedActions, type) => ({
    ...modifiedActions,
    [type]: actions[type](dispatch, state),
  }), {});
