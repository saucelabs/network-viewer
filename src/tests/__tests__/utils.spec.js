/* eslint no-useless-escape: 0 */

import * as utils from '../../utils';
import networkDataMock from '../__fixtures__/network.json';
import preparedMockData from '../__fixtures__/preparedData';

describe('utils', () => {
  it('getUrlInfo', () => {
    const url = 'https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2';
    expect(utils.getUrlInfo(url)).toMatchSnapshot();
  });

  it('parseSize', () => {
    const { response } = networkDataMock.log.entries[0];
    expect(utils.parseSize(response)).toMatchSnapshot();
  });

  it('getContentType', () => {
    const { headers } = networkDataMock.log.entries[0].response;
    expect(utils.getContentType(headers)).toMatchSnapshot();
  });

  it('getTimings', () => {
    const firstEntryTime = networkDataMock.log.entries[0].startedDateTime;
    expect(utils.getTimings(networkDataMock.log.entries[3], firstEntryTime)).toMatchSnapshot();
  });

  it('prepareViewerData', () => {
    const { entries } = networkDataMock.log;
    expect(utils.prepareViewerData(entries)).toMatchSnapshot();
  });

  it('sortBy', () => {
    expect(utils.sortBy(preparedMockData, 'size').toJS()).toMatchSnapshot();
  });

  it('filterData', () => {
    expect(utils.filterData({
      data: preparedMockData,
      search: { name: 'url', value: 'e96c15f68c68' },
      filter: {},
    }).toJS()).toMatchSnapshot();
    expect(utils.filterData({
      data: preparedMockData,
      filter: { key: 'type', value: 'html' },
      search: {},
    }).toJS()).toMatchSnapshot();
  });

  it('getContent', () => {
    expect(utils.getContent({
      mimeType: 'application/json',
      text: '{\n  \"foo\": \"bar\"\n}',
    })).toMatchSnapshot();
    expect(utils.getContent({ content: 'cool' })).toMatchSnapshot();
  });
});
