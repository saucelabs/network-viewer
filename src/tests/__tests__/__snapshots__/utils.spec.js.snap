// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils filterData 1`] = `
Array [
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
    "index": 2,
    "method": "GET",
    "size": 33.27,
    "startedDateTime": "2019-10-14T08:47:11.641Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 13.034000003244728,
      "blocked": 14.968000003244727,
      "connect": -1,
      "dns": -1,
      "receive": 1.9629999878816307,
      "send": 0.40400000000000014,
      "ssl": -1,
      "startTime": 49,
      "wait": 35.81200000759028,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
  },
]
`;

exports[`utils filterData 2`] = `
Array [
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "slice",
    "index": 0,
    "method": "GET",
    "size": 135.04,
    "startedDateTime": "2019-10-14T08:47:11.592Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 4.382999992230907,
      "blocked": 5.496999992230907,
      "connect": -1,
      "dns": -1,
      "receive": 3.252000024076551,
      "send": 0.2819999999999998,
      "ssl": -1,
      "startTime": 0,
      "wait": 26.03899999633059,
    },
    "type": "document",
    "url": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
    "index": 1,
    "method": "GET",
    "size": 33.02,
    "startedDateTime": "2019-10-14T08:47:11.641Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 12.528999999631196,
      "blocked": 13.952999999631196,
      "connect": -1,
      "dns": -1,
      "receive": 7.124000025214627,
      "send": 0.2370000000000001,
      "ssl": -1,
      "startTime": 49,
      "wait": 18.80099998585507,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
    "index": 2,
    "method": "GET",
    "size": 33.27,
    "startedDateTime": "2019-10-14T08:47:11.641Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 13.034000003244728,
      "blocked": 14.968000003244727,
      "connect": -1,
      "dns": -1,
      "receive": 1.9629999878816307,
      "send": 0.40400000000000014,
      "ssl": -1,
      "startTime": 49,
      "wait": 35.81200000759028,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
  },
  Object {
    "body": ".logo{display:block;grid-area:I;width:200px;height:44px;margin-right:24px}",
    "domain": "developer.mozilla.org",
    "filename": "react-header.2a659f6ad280.css",
    "index": 3,
    "method": "GET",
    "size": 2.76,
    "startedDateTime": "2019-10-14T08:47:11.643Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 11.349999986123294,
      "blocked": 13.026999986123293,
      "connect": -1,
      "dns": -1,
      "receive": 1.951000012923032,
      "send": 0.21299999999999986,
      "ssl": -1,
      "startTime": 51,
      "wait": 36.02600000113249,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-header.2a659f6ad280.css",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "react-mdn.c66d8d89d98e.css",
    "index": 4,
    "method": "GET",
    "size": 142.88,
    "startedDateTime": "2019-10-14T08:47:11.644Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 11.250000010477379,
      "blocked": 13.423000010477379,
      "connect": -1,
      "dns": -1,
      "receive": 1.2719999940600246,
      "send": 0.5390000000000001,
      "ssl": -1,
      "startTime": 52,
      "wait": 37.96100001042895,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-mdn.c66d8d89d98e.css",
  },
]
`;

exports[`utils getContent 1`] = `"{\\"foo\\":\\"bar\\"}"`;

exports[`utils getContent 2`] = `undefined`;

exports[`utils getContentType 1`] = `"html"`;

exports[`utils getTimings 1`] = `
Object {
  "_blocked_queueing": 11.349999986123294,
  "blocked": 13.026999986123293,
  "connect": -1,
  "dns": -1,
  "receive": 1.951000012923032,
  "send": 0.21299999999999986,
  "ssl": -1,
  "startTime": 51,
  "wait": 36.02600000113249,
}
`;

exports[`utils getUrlInfo 1`] = `
Object {
  "domain": "developer.mozilla.org",
  "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
  "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
}
`;

exports[`utils parseSize 1`] = `135.04`;

exports[`utils prepareViewerData 1`] = `
Object {
  "data": Array [
    Object {
      "body": "",
      "domain": "developer.mozilla.org",
      "filename": "slice",
      "index": 0,
      "method": "GET",
      "size": 135.04,
      "startedDateTime": "2019-10-14T08:47:11.592Z",
      "status": 200,
      "timings": Object {
        "_blocked_queueing": 4.382999992230907,
        "blocked": 5.496999992230907,
        "connect": -1,
        "dns": -1,
        "receive": 3.252000024076551,
        "send": 0.2819999999999998,
        "ssl": -1,
        "startTime": 0,
        "wait": 26.03899999633059,
      },
      "type": "document",
      "url": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice",
    },
    Object {
      "body": "",
      "domain": "developer.mozilla.org",
      "filename": "ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
      "index": 1,
      "method": "GET",
      "size": 33.02,
      "startedDateTime": "2019-10-14T08:47:11.641Z",
      "status": 200,
      "timings": Object {
        "_blocked_queueing": 12.528999999631196,
        "blocked": 13.952999999631196,
        "connect": -1,
        "dns": -1,
        "receive": 7.124000025214627,
        "send": 0.2370000000000001,
        "ssl": -1,
        "startTime": 49,
        "wait": 18.80099998585507,
      },
      "type": "font",
      "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
    },
    Object {
      "body": "",
      "domain": "developer.mozilla.org",
      "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
      "index": 2,
      "method": "GET",
      "size": 33.27,
      "startedDateTime": "2019-10-14T08:47:11.641Z",
      "status": 200,
      "timings": Object {
        "_blocked_queueing": 13.034000003244728,
        "blocked": 14.968000003244727,
        "connect": -1,
        "dns": -1,
        "receive": 1.9629999878816307,
        "send": 0.40400000000000014,
        "ssl": -1,
        "startTime": 49,
        "wait": 35.81200000759028,
      },
      "type": "font",
      "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
    },
    Object {
      "body": ".logo{display:block;grid-area:I;width:200px;height:44px;margin-right:24px}",
      "domain": "developer.mozilla.org",
      "filename": "react-header.2a659f6ad280.css",
      "index": 3,
      "method": "GET",
      "size": 2.76,
      "startedDateTime": "2019-10-14T08:47:11.643Z",
      "status": 200,
      "timings": Object {
        "_blocked_queueing": 11.349999986123294,
        "blocked": 13.026999986123293,
        "connect": -1,
        "dns": -1,
        "receive": 1.951000012923032,
        "send": 0.21299999999999986,
        "ssl": -1,
        "startTime": 51,
        "wait": 36.02600000113249,
      },
      "type": "stylesheet",
      "url": "https://developer.mozilla.org/static/build/styles/react-header.2a659f6ad280.css",
    },
    Object {
      "body": "",
      "domain": "developer.mozilla.org",
      "filename": "react-mdn.c66d8d89d98e.css",
      "index": 4,
      "method": "GET",
      "size": 142.88,
      "startedDateTime": "2019-10-14T08:47:11.644Z",
      "status": 200,
      "timings": Object {
        "_blocked_queueing": 11.250000010477379,
        "blocked": 13.423000010477379,
        "connect": -1,
        "dns": -1,
        "receive": 1.2719999940600246,
        "send": 0.5390000000000001,
        "ssl": -1,
        "startTime": 52,
        "wait": 37.96100001042895,
      },
      "type": "stylesheet",
      "url": "https://developer.mozilla.org/static/build/styles/react-mdn.c66d8d89d98e.css",
    },
  ],
  "totalNetworkTime": 53.271999994060025,
}
`;

exports[`utils sortBy 1`] = `
Array [
  Object {
    "body": ".logo{display:block;grid-area:I;width:200px;height:44px;margin-right:24px}",
    "domain": "developer.mozilla.org",
    "filename": "react-header.2a659f6ad280.css",
    "index": 3,
    "method": "GET",
    "size": 2.76,
    "startedDateTime": "2019-10-14T08:47:11.643Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 11.349999986123294,
      "blocked": 13.026999986123293,
      "connect": -1,
      "dns": -1,
      "receive": 1.951000012923032,
      "send": 0.21299999999999986,
      "ssl": -1,
      "startTime": 51,
      "wait": 36.02600000113249,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-header.2a659f6ad280.css",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
    "index": 1,
    "method": "GET",
    "size": 33.02,
    "startedDateTime": "2019-10-14T08:47:11.641Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 12.528999999631196,
      "blocked": 13.952999999631196,
      "connect": -1,
      "dns": -1,
      "receive": 7.124000025214627,
      "send": 0.2370000000000001,
      "ssl": -1,
      "startTime": 49,
      "wait": 18.80099998585507,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
    "index": 2,
    "method": "GET",
    "size": 33.27,
    "startedDateTime": "2019-10-14T08:47:11.641Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 13.034000003244728,
      "blocked": 14.968000003244727,
      "connect": -1,
      "dns": -1,
      "receive": 1.9629999878816307,
      "send": 0.40400000000000014,
      "ssl": -1,
      "startTime": 49,
      "wait": 35.81200000759028,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "slice",
    "index": 0,
    "method": "GET",
    "size": 135.04,
    "startedDateTime": "2019-10-14T08:47:11.592Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 4.382999992230907,
      "blocked": 5.496999992230907,
      "connect": -1,
      "dns": -1,
      "receive": 3.252000024076551,
      "send": 0.2819999999999998,
      "ssl": -1,
      "startTime": 0,
      "wait": 26.03899999633059,
    },
    "type": "document",
    "url": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "react-mdn.c66d8d89d98e.css",
    "index": 4,
    "method": "GET",
    "size": 142.88,
    "startedDateTime": "2019-10-14T08:47:11.644Z",
    "status": 200,
    "timings": Object {
      "_blocked_queueing": 11.250000010477379,
      "blocked": 13.423000010477379,
      "connect": -1,
      "dns": -1,
      "receive": 1.2719999940600246,
      "send": 0.5390000000000001,
      "ssl": -1,
      "startTime": 52,
      "wait": 37.96100001042895,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-mdn.c66d8d89d98e.css",
  },
]
`;
